Core/GDCore/IDE/WholeProjectRefactorer.cpp

```diff
@@ -321,8 +330,83 @@ void WholeProjectRefactorer::ApplyRefactoringForVariablesContainer(
+                                            changeset.typeChangedVariableNames,
+                                            variablesContainer);
+  gd::ProjectBrowserHelper::ExposeProjectEvents(
+      project, eventsVariableInstructionTypeSwitcher);
+}
+
+void WholeProjectRefactorer::ApplyRefactoringForGroupVariablesContainer(
+    gd::Project &project, gd::ObjectsContainer &globalObjectsContainer,
+    gd::ObjectsContainer &objectsContainer,
+    const gd::VariablesContainer &groupVariablesContainer,
+    const gd::ObjectGroup &objectGroup,
+    const gd::VariablesChangeset &changeset,
+    const gd::SerializerElement &originalSerializedVariables) {
+
+  // While we support refactoring that would remove all references (actions, conditions...)
+  // it's both a bit dangerous for the user and we would need to show the user what
+  // will be removed before doing so. For now, just clear the removed variables so they don't
+  // trigger any refactoring.
+  std::unordered_set<gd::String> removedVariableNames;
+
+  // Rename variables in events for the group objects.
+  for (const gd::String &objectName : objectGroup.GetAllObjectsNames()) {
+    const bool hasObject = objectsContainer.HasObjectNamed(objectName);
+    if (!hasObject && !globalObjectsContainer.HasObjectNamed(objectName)) {
+      continue;
+    }
+    auto &object = hasObject ? objectsContainer.GetObject(objectName)
+                             : globalObjectsContainer.GetObject(objectName);
+    auto &variablesContainer = object.GetVariables();
+
+    gd::EventsVariableReplacer eventsVariableReplacer(
+        project.GetCurrentPlatform(), changeset,
+        removedVariableNames, variablesContainer);
+    gd::ProjectBrowserHelper::ExposeProjectEvents(project,
+                                                  eventsVariableReplacer);
+  }
+
+  // Rename variables in events for the group.
+  gd::EventsVariableReplacer eventsVariableReplacer(
+      project.GetCurrentPlatform(), changeset, removedVariableNames,
+      objectGroup.GetName());
+  gd::ProjectBrowserHelper::ExposeProjectEvents(project,
+                                                eventsVariableReplacer);
+
+  // Apply changes to objects.
+  gd::GroupVariableHelper::FillMissingGroupVariablesToObjects(
+      globalObjectsContainer,
+      objectsContainer,
+      objectGroup,
+      originalSerializedVariables);
+  gd::GroupVariableHelper::ApplyChangesToObjects(
+      globalObjectsContainer, objectsContainer, groupVariablesContainer,
+      objectGroup, changeset);
+
+  // Switch types of instructions for the group objects.
+  for (const gd::String &objectName : objectGroup.GetAllObjectsNames()) {
+    const bool hasObject = objectsContainer.HasObjectNamed(objectName);
+    if (!hasObject && !globalObjectsContainer.HasObjectNamed(objectName)) {
+      continue;
+    }
+    auto &object = hasObject ? objectsContainer.GetObject(objectName)
+                             : globalObjectsContainer.GetObject(objectName);
```

Comment:
Could almost be a helper function (with a comment that it should probably one day be an ObjectsContainersList. Passing two ObjectsContainer is a smell that it's not future proof).

Reply:
I agree, but `ObjectsContainersList` only grant read-only access. There is maybe a distinction to do between moving the context and doing changes in the context.
Maybe moving the context should be done by creating a new context. A `const ObjectsContainersList` would mean that no changes can be done to the objects.

Reply:
Ok, we'll see later how we pass "not just read only" list of objects containers :)